File: ecolink-back/src/main/java/com/ecolink/model/Schedule.java
package com.ecolink.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "schedules", indexes = {
    @Index(name = "idx_schedule_userId", columnList = "user_id"),
    @Index(name = "idx_schedule_status", columnList = "status"),
    @Index(name = "idx_schedule_date", columnList = "scheduledDate")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Schedule {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 255)
    private String wasteType;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String address;

    @Column(nullable = false)
    private Double latitude;

    @Column(nullable = false)
    private Double longitude;

    @Column(nullable = false)
    private LocalDateTime scheduledDate;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ScheduleStatus status;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public enum ScheduleStatus {
        PENDING, CONFIRMED, COMPLETED, CANCELLED
    }

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        if (this.createdAt == null) this.createdAt = now;
        if (this.updatedAt == null) this.updatedAt = now;
        if (this.status == null)    this.status    = ScheduleStatus.PENDING;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}

File: flutter_application_1/lib/models/schedule.dart
enum ScheduleStatus { pending, confirmed, completed, cancelled }

ScheduleStatus scheduleStatusFromString(String? value) {
  switch (value) {
    case 'CONFIRMED':
      return ScheduleStatus.confirmed;
    case 'COMPLETED':
      return ScheduleStatus.completed;
    case 'CANCELLED':
      return ScheduleStatus.cancelled;
    case 'PENDING':
    default:
      return ScheduleStatus.pending;
  }
}

class Schedule {
  final int? id;
  final String wasteType;
  final String address;
  final double latitude;
  final double longitude;
  final DateTime scheduledDate;
  final String? description;
  final ScheduleStatus status;
  final DateTime? createdAt;

  Schedule({
    this.id,
    required this.wasteType,
    required this.address,
    required this.latitude,
    required this.longitude,
    required this.scheduledDate,
    this.description,
    this.status = ScheduleStatus.pending,
    this.createdAt,
  });

  factory Schedule.fromJson(Map<String, dynamic> json) {
    return Schedule(
      id: json['id'] as int?,
      wasteType: json['wasteType'] as String? ?? '',
      address: json['address'] as String? ?? '',
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0,
      scheduledDate: DateTime.tryParse(json['scheduledDate'] as String? ?? '') ??
          DateTime.now(),
      description: json['description'] as String?,
      status: scheduleStatusFromString(json['status'] as String?),
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toCreateJson(Map<String, dynamic> userRef) {
    return {
      'wasteType': wasteType,
      'address': address,
      'latitude': latitude,
      'longitude': longitude,
      'scheduledDate': scheduledDate.toIso8601String(),
      'description': description,
      'user': userRef,
    };
  }
}
